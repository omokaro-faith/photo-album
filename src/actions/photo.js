import axios from 'axios';

export const setPhotos = photos => ({
  type: 'GET_PHOTOS',
  photos,
});

export const setPhotoError = () => ({
  type: 'GET_PHOTOS_ERROR',
  message: true
});

export const setPhotosLength = totalPhotos => ({
  type: 'GET_ALL_PHOTOS',
  totalPhotos,
});

export const fetchAllPhotosError = () => ({
  type: 'GET_ALL_PHOTOS_ERROR',
  errorMessage: true
});

export const fetchAllPhotos = albumId => async (dispatch) => {
  try {
    const photos = await axios.get(`https://jsonplaceholder.typicode.com/photos?albumId=${albumId}`);
    return dispatch(setPhotosLength(photos.data.length));
  } catch (err) {
    return(dispatch(fetchAllPhotosError()));
  }
}

export const getPhotos = (albumId, start= 0 , limit) => async (dispatch) => {
  try {
    const photos = await axios.get(`https://jsonplaceholder.typicode.com/photos?albumId=${albumId}&_start=${start}&_limit=${limit}`);
    return dispatch(setPhotos(photos.data));
  } catch (err) {
    return(dispatch(setPhotoError()));
  }
};
