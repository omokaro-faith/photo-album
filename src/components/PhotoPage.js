import React from 'react';
const PhotoPage = () => (
  <header>
    <h1>Faith's PhotoAlbum</h1>
  </header>
)

 export default PhotoPage; 