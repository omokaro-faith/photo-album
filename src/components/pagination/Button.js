import React from 'react';
import { PropTypes } from 'prop-types';

const Button = ({id, handleClick, pageNumber}) => (
  <li
    id={id}
    onClick={handleClick}
  >
    {pageNumber}
  </li>
)

Button.propTypes = {
  id: PropTypes.number.isRequired,
  handleClick: PropTypes.func.isRequired,
  pageNumber: PropTypes.number.isRequired,
}



export default Button;
