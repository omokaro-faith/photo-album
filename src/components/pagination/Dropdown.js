import React from 'react';
import { PropTypes } from 'prop-types';

const Dropdown = ({handleChange}) => (
  <select onChange={handleChange}>
    <option>10</option>
    <option>20</option>
    <option>30</option>
    <option>50</option>
  </select>
)

Dropdown.propTypes = {
  handleChange: PropTypes.func.isRequired,
}



export default Dropdown;
