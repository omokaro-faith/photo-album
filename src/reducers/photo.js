const initialState = {isLoading: true}

 export default (state= initialState, action) => {
 switch (action.type) {
   case 'GET_PHOTO':
   const { users } = action
     return {...state, users, isLoading: false};
   default:
     return initialState;
 }
}