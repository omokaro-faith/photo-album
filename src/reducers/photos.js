const initialState = {}

 export default (state= initialState, action) => {
 switch (action.type) {
   case 'GET_PHOTOS':
   const { photos } = action;
     return {...state, photos};
    case 'GET_ALL_PHOTOS':
      const { totalPhotos } = action;
    return {...state, totalPhotos};
   case 'GET_PHOTOS_ERROR':
   const { message } = action; 
     return { ...state, message };
   case 'GET_ALL_PHOTOS_ERROR':
      const { errorMessage } = action; 
      return { ...state, errorMessage}
   default:
     return state;
 }
}
