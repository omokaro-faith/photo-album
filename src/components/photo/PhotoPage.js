import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Dropdown from '../pagination/Dropdown';
import Buttons from '../pagination/Button';
import Modal from '../modal/Modal';
import { getPhotos }  from '../../actions/photo';
import Header from '../header/Header';
import { getPageNumbers, getCurrentItems } from '../../utils/utils';

export class PhotoPage extends Component {
  constructor(props) {
    super(props);
    this.state = {
      show: false,
      modalContent: {},
      currentPage: 1,
      itemsPerPage: 10,
    };
    this.handleChange = this.handleChange.bind(this);
    this.showModal = this.showModal.bind(this);
    this.handleClick = this.handleClick.bind(this);
  };

  static propTypes = {
    getPhotos: PropTypes.func.isRequired,
    photos: PropTypes.array.isRequired,
    match: PropTypes.shape({
      params: PropTypes.shape({
        albumOwner: PropTypes.string.isRequired,
        albumTitle: PropTypes.string.isRequired
      })
    }).isRequired
  };

  componentDidMount() {
    const { match: { params }, getPhotos } = this.props;
    getPhotos(undefined, params.albumId)
  }

  handleChange(event) {
    const { match: { params }, getPhotos } = this.props;
    getPhotos(event.target.value, params.albumId);
    this.setState({
      currentPage: 1,
    });
  }

  handleClick(event) {
    this.setState({
      currentPage: Number(event.target.id)
    });
  }

  showModal = ({ content }) => {
    this.setState({ 
      show: true,
      modalContent: {
        picture: content.url,
        title: content.title,
      }
     });
  };

  hideModal = () => {
    this.setState({ 
      show: false,
      modalContent: {},
     });
  };

  render() {
    const { photos, match: { params: { albumOwner, albumTitle } } } = this.props;
    const { show, modalContent, currentPage, itemsPerPage } = this.state;
    const renderPhotos = getCurrentItems(currentPage, itemsPerPage, photos).map((photo, index) => 
    (
      <div key={index} className="grid__item" onClick={() => this.showModal({ content: photo })}>
        <div className="grid__img">
            <img src={`${photo.thumbnailUrl}`} alt='' title='' className='grid__img--item'/>
        </div>
        <h5>Title: {photo.title}</h5>
      </div>
    ));

    const renderPageNumbers = getPageNumbers(photos, itemsPerPage).map((number) => (
      <Buttons 
       key={number}
       id={number}
       handleClick={this.handleClick}
       pageNumber={number}
      />
   ))


    return (
      <section className="grid">
        <Header totalItems={photos.length} pageName='Photos'/>
        <div className='dropdown-button__wrapper'>
         <Dropdown handleChange={this.handleChange} itemsLength={photos.length}/>
         {renderPageNumbers}
        </div>
        { photos.length >= 1 ? 
          <Fragment>
            <h3>Owner: {albumOwner}</h3>
            <h3>Album Title: {albumTitle}</h3>
            <div className="grid__container">
            { renderPhotos }
            </div>
          </Fragment>
        : <div className="loader"></div> }
        <Modal show={show} handleClose={this.hideModal}>
            <div className='img-content'>
              <img src={`${modalContent.picture}`} alt='' title=''/>
            </div>
            <div>
              <h5>Owner: <span>{albumOwner}</span></h5>
              <h5>Title: {modalContent.title}</h5>
            </div>
        </Modal>
      </section> 
    )
  }
}

PhotoPage.defaultProps = {
  photos: [],
};

const mapDispatchToProps = (dispatch) => ({
  getPhotos: (limit, albumId) => dispatch(getPhotos(limit, albumId)),
});

 const mapStateToProps = (state) => ({
  photos: state.photos.photos,
});



export default connect(mapStateToProps, mapDispatchToProps)(PhotoPage);
