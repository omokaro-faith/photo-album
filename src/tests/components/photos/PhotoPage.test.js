import React from 'react';
import { shallow } from 'enzyme';
import { PhotoPage } from '../../../components/photo/PhotoPage';
import Dropdown from '../../../components/pagination/Dropdown';
import Button from '../../../components/pagination/Button';
import photos from '../../fixtures/photos';

describe('PhotoPage', () => {
  let props;
  let wrapper;

  beforeEach(() => {
    props = {
      photos: [],
      getPhotos: jest.fn(),
      fetchAllPhotos: jest.fn(),
      totalPhotos: 10,
      match: {
        params: {
          albumOwner: 'Leonardo Da vinci',
          albumTitle: 'Angel & Demos',
        }
      }
    };
    wrapper = shallow(<PhotoPage {...props}/>);
  });

  afterEach(() => { wrapper = null; });

  
  describe('renders', () => {
    it('should exist', () => {
      expect(wrapper.exists()).toBe(true);
    });

    it('should not have any regressions', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('sets photos', () => {
      wrapper.setState({
       photos: photos
      });

      const result = {
        modalContent:  {},
        photos: [
          {
           id: 1,
           thumbnailUrl: 'http://dummypath',
           url: 'http://dummypath',
         },
          {
           id: 2,
           thumbnailUrl: 'http://dummypath',
           url: 'http://dummypath',
         },
          {
           id: 3,
           thumbnailUrl: 'http://dummypath',
           url: 'http://dummypath',
         },
          {
           id: 4,
           thumbnailUrl: 'http://dummypath',
           url: 'http://dummypath',
         },
          {
           id: 5,
           thumbnailUrl: 'http://dummypath',
           url: 'http://dummypath',
         },
        ],
        show: false,
        currentPage: 1,
        itemsPerpage: 20,
      };

      expect(wrapper.state()).toEqual(result);
    });

    it('displays the loader', () => {
      const loader = wrapper.find('.loader');
      expect(loader.length).toEqual(1);
    });

    it('displays the dropdown component', () => {
      const select = wrapper.find(Dropdown);
      expect(select.length).toEqual(1);
    });

    it('displays the button component', () => {
      const button = wrapper.find(Button);
      expect(button.length).toEqual(1);
    });
  });
})