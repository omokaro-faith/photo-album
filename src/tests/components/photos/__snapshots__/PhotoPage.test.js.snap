// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhotoPage renders should not have any regressions 1`] = `ShallowWrapper {}`;
